lib/tapp.hoon

```diff
@@ -225,24 +249,66 @@
       =.  waiting  (~(put to waiting) %init ~)
       start-async
     ::
-    =/  old  ((soft tapp-state) u.old-state)
+    =/  old   ((soft tapp-state) u.old-state)
     ?~  old
       ~&  [%tapp-reset dap.bowl]
+      ::  XX may break contracts!
+      ::  XX if active clam contracts only to abort transaction?
+      ::
       `this-tapp
+    ::
+    ::  because the clam replaces the active continuation with
+    ::  the bunt of its mold, we must fail the transaction
+    ::
     ~&  [%tapp-loaded dap.bowl]
-    `this-tapp(+<+ u.old)
+    =.  +<+.this-tapp  u.old
+    ?^  active
+      =.  waiting  (~(put to waiting) (need ~(top to waiting)))
+      (oob-fail-async %reset-restart ~)
+    `this-tapp
   ::
   ::  Start a command
   ::
   ++  poke
-    |=  =in-poke-data
+    |=  =tapp-in-poke-data
     ^-  (quip move _this-tapp)
-    =.  waiting  (~(put to waiting) %poke in-poke-data)
+    ?:  ?=(tapp-admin-in-poke-data tapp-in-poke-data)
+      ?~  active
+        ~&  [%tapp-admin-idle dap.bowl]
+        `this-tapp
+      ?-  tapp-admin.tapp-in-poke-data
+          %cancel
+        (oob-fail-async %tapp-admin-cancel ~)
+      ::
+          %restart
+        =.  waiting  (~(put to waiting) (need ~(top to waiting)))
+        (oob-fail-async %tapp-admin-restart ~)
+      ==
+    ::
+    =.  waiting  (~(put to waiting) %poke tapp-in-poke-data)
     ?^  active
       ~&  [%waiting-until-current-async-finishes waiting]
       `this-tapp
     start-async
   ::
+  ::  Receive acknowledgement of outgoing poke
+  ::
+  ::    XX these can be distinguished by dock, but, without a wire or
+  ::    some alternative, that's not very useful if you poke the same
+  ::    dock multiple times. %poke wires are not currently exposed ...
+  ::    Wat do?
```

Comment:
Wires are definitely the right way to handle this.  I guess we should let the user specify a wire when poking?

Reply:
Like, so it ends up in the bone?

Reply:
Not in the bone (at least, not from this app's perspective).  This is a `%coup` response to a poke, so whatever wire we sent on the poke comes back to us as our first argument.  We can then switch on that if we want.